### PR TITLE
sigids: derive a sigid from a sig payload (reconstructed)

### PR DESCRIPTION
**Pull request by maxtaco**
- this is largely straightforward as you would expect, but clients at versions <1.0.16 were all buggy
- some clients at 1.0.16 (linux nightlies) are fixed, and some aren't
- keep in mind the client version is signed over, so it's safe to switch on it in the code
- figure out all signatures generated with 1.0.16, and make a memory-optimized lookup table to indicate the broken ones
- good testing around the different cases in our derivation routines

### CODE REVIEW

File: go/sigid/sigid.go [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+// 1.0.16 compute SigIDs the right way, and some do it the wrong way,
+// (likely because Linux nightlies were going out with the 1.0.16 version).
+//
+// There are 77331 total signatures that were generated via 1.0.16, of which
+// 16555 where generated the wrong way (using the <= 1.0.16 method). In data.go,
+// we have prefixes of wrong/legacy signature IDs, that need to be fixed. To save
+// binary space, we only include in data.go the shortest possible prefix of the
+// wrong/legacy signature IDs that don't collide with prefixes from the group of
+// correct signatures. So the idea is that if we compute a hash the right way,
+// then find its prefix in data.go, then we need to recompute it the wrong way,
```
Comment (heronhaye) [inline go/sigid/sigid.go]: confused here. Does "Right" mean modern and "Wrong" mean legacy (due to the `hash` bug)? There's no security issue with the "wrong" one right?

So does `data.go` contain what's in the merkle tree/db? Or is it the modern version for all of them?

File: go/sigid/data/mkdata.iced [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+class Runner
+
+  constructor : () ->
+    map = []
+
+  read : ({input}, cb) ->
+    esc = make_esc cb, "read"
+    gets = (new Gets input).run()
+    h = {} # Sigs that start with 'h' are the fixed/new ones
```
Comment (heronhaye) [inline go/sigid/data/mkdata.iced]: maybe link to the pr and comment that his has to do with msgpack encoding of different length structures?

Comment (heronhaye) [reply go/sigid/data/mkdata.iced]: what does "-" mean in sigids-1.0.16? neither g nor h? what caused those sigs?

File: go/kbcrypto/kbpackets.go [hunk 3]
```diff
@@ -170,23 +178,24 @@ type keybasePacket struct {
 	Version PacketVersion      `codec:"version"`
 }
 
-func newKeybasePacket(body Packetable) (*keybasePacket, error) {
+func newKeybasePacket(body Packetable, doHash bool) (*keybasePacket, error) {
```
[the remark this answers is not in the document]
Comment (heronhaye) [reply go/kbcrypto/kbpackets.go]: and/or change variable name to "isModern"

File: go/sigid/sigid.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+//
+// Background:
+//
+// The commit that fixes the 2016 bug is:
+//
+//     https://github.com/keybase/client/commit/ca023e2d6f9192fd8e923f3113ae4a11cda8b53a
+//
+// So code on opposite sides of this bug will generate SigIDs in two
+// different ways.
+func isMaybeModernSigIDMakerModern(sigID keybase1.SigID) bool {
```
[the remark this answers is not in the document]
Comment (maxtaco) [reply go/sigid/sigid.go]: i thought go doesn't like underscores in function names?

**Comment by maxtaco** [comment]
There's a ticket for that, but we can do that before the next PR is merged, sure.

File: go/sigid/sigid.go (hunk 1 above)
Comment (maxtaco) [reply go/sigid/sigid.go]: modern version of them.

File: go/sigid/data/mkdata.iced (hunk 2 above)
Comment (maxtaco) [reply go/sigid/data/mkdata.iced]: PGP

**Pull request merged**